ports/zephyr/zephyr_dynamic_pinctrl.c

```diff
@@ -0,0 +1,250 @@
+    if (!mp_obj_is_type(obj, &zephyr_pinctrl_type)) {
+        mp_raise_ValueError(MP_ERROR_TEXT("Argument is not of type zephyr.Pinctrl"));
+    }
+    zephyr_pinctrl_obj_t *this = MP_OBJ_TO_PTR(obj);
+    return this->pinctrl;
+}
+
+static void zephyr_pinctrl_print(const mp_print_t *print, mp_obj_t self_in, mp_print_kind_t kind) {
+    zephyr_pinctrl_obj_t *self = MP_OBJ_TO_PTR(self_in);
+    mp_printf(print, "Pinctrl(%d, %d, pu=%s, pd=%s, schmitt_enable=%s, \
+out_enable=%s, in_enable=%s, slew=%d, drive=%d)",
+        mp_dynamic_pinctrl_get_pin(self->pinmux),
+        mp_dynamic_pinctrl_get_function(self->pinmux),
+        self->bias_pull_up ? "True" : "False",
+        self->bias_pull_down ? "True" : "False",
+        self->input_schmitt_enable ? "True" : "False",
+        self->output_enable ? "True" : "False",
+        self->input_enable ? "True" : "False",
+        self->slew_rate,
+        self->drive_strength
+        );
+}
+
+static void zephyr_pinctrl_make_new_kw_args(zephyr_pinctrl_obj_t *self,
+    size_t n_args, const mp_obj_t *pos_args, mp_map_t *kw_args) {
+    enum { ARG_pu, ARG_pd, ARG_schmitt_enable, ARG_out_enable, ARG_in_enable, ARG_slew, ARG_drive };
+
+    static const mp_arg_t allowed_args[] = {
+        { MP_QSTR_pu, MP_ARG_KW_ONLY | MP_ARG_BOOL, {.u_bool = false } },
+        { MP_QSTR_pd, MP_ARG_KW_ONLY | MP_ARG_BOOL, {.u_bool = false } },
+        { MP_QSTR_schmitt_enable, MP_ARG_KW_ONLY | MP_ARG_BOOL, {.u_bool = false } },
+        { MP_QSTR_out_enable, MP_ARG_KW_ONLY | MP_ARG_BOOL, {.u_bool = false } },
+        { MP_QSTR_in_enable, MP_ARG_KW_ONLY | MP_ARG_BOOL, {.u_bool = false } },
+        { MP_QSTR_slew, MP_ARG_KW_ONLY | MP_ARG_INT, {.u_int = 0 } },
+        { MP_QSTR_drive, MP_ARG_KW_ONLY | MP_ARG_INT, {.u_int = 2 } },
+    };
+
+    mp_arg_val_t args[MP_ARRAY_SIZE(allowed_args)];
+    mp_arg_parse_all(n_args, pos_args, kw_args, MP_ARRAY_SIZE(allowed_args), allowed_args, args);
+
+    self->bias_pull_up = args[ARG_pu].u_bool;
+    self->bias_pull_down = args[ARG_pd].u_bool;
+    self->input_schmitt_enable = args[ARG_schmitt_enable].u_bool;
+    self->output_enable = args[ARG_out_enable].u_bool;
+    self->input_enable = args[ARG_in_enable].u_bool;
+    self->slew_rate = args[ARG_slew].u_int;
+    self->drive_strength = args[ARG_drive].u_int;
+
+    self->pinctrl = mp_dynamic_pinctrl_make_pinctrl_soc_pin_t(self->pinmux,
+        args[ARG_pu].u_bool,
+        args[ARG_pd].u_bool,
+        args[ARG_schmitt_enable].u_bool,
+        args[ARG_out_enable].u_bool,
+        args[ARG_in_enable].u_bool,
+        args[ARG_slew].u_int,
+        args[ARG_drive].u_int);
+}
+
+static mp_obj_t zephyr_pinctrl_make_new(const mp_obj_type_t *type,
+    size_t n_args, size_t n_kw, const mp_obj_t *args) {
+    #ifdef CONFIG_MICROPY_DYNAMIC_PINCTRL_GENERATE_PINMUX
+    size_t index = 0;
+    #endif
+    mp_arg_check_num(n_args, n_kw, 1, MP_OBJ_FUN_ARGS_MAX, true);
+    bool arbitrary = true;
```

Comment:
How about some hint for early users seeking to understand the feature?
```c
// For SoCs where any pin can be used for any function, both passed as integers
```
Not sure my wording's best.

Reply:
That's not the goal there. The goal is to provide the ability to use the dynamic pinctrl even on platforms without pinmux parsing, and also as a alternative to pinmux names if the user knows the function numbers (or has added a a database for that in their code).

Reply:
Some way to get direct low-level access without the strings then... perfect for when `CONFIG_MICROPY_DYNAMIC_PINCTRL_GENERATE_PINMUX=n`.